sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/BulkExecutor.java

```diff
@@ -150,13 +155,78 @@ public BulkExecutor(CosmosAsyncContainer container,
         // To make sure we flush the buffers at least every maxMicroBatchIntervalInMs we start a timer
         // that will trigger artificial ItemOperations that are only used to flush the buffers (and will be
         // filtered out before sending requests to the backend)
-        this.executorService = Executors.newSingleThreadScheduledExecutor(
-                new CosmosDaemonThreadFactory("BulkExecutor-" + instanceCount.incrementAndGet()));
+        // this.executorService = Executors.newSingleThreadScheduledExecutor(
+        //        new CosmosDaemonThreadFactory("BulkExecutor-" + instanceCount.incrementAndGet()));
+        this.executorService = new ScheduledThreadPoolExecutor(
+            1,
+            new CosmosDaemonThreadFactory(identifier));
+        this.executorService.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
+        this.executorService.setRemoveOnCancelPolicy(true);
         this.scheduledFutureForFlush = this.executorService.scheduleWithFixedDelay(
             this::onFlush,
             this.maxMicroBatchIntervalInMs,
             this.maxMicroBatchIntervalInMs,
             TimeUnit.MILLISECONDS);
+
+        logger.debug("Instantiated BulkExecutor, Context: {}",
+            this.operationContextText);
+    }
+
+    @Override
+    public void dispose() {
+        if (this.isDisposed.compareAndSet(false, true)) {
+            long totalCountSnapshot = totalCount.get();
+            if (totalCountSnapshot == 0) {
```

Comment:
question: why need to differentiate here? why not call completeAllSinks() in both case?

Reply:
completeAllSinks attempts to complete the sink - which indicates successful completion - which we shouldn't when there is work left and cancellation happens